### DISCUSSION TITLE
Support of Spline for Glue AWS

### DISCUSSION THREAD

**Discussion by zacayd** [AbsaOSS/spline-spark-agent#647] [Q&A]
Hi
does Spline support AWS Glues that runs jobs on Spark?
if so do you have any manual how to configure it?
Zacay

**Comment by cerveada** [comment]
> does Spline support AWS Glues that runs jobs on Spark?

The Spline developement and testing is done primarly on [Apache Spark](https://spark.apache.org/) we try to fix issues with other flavours of Spark as well, but we don't have resources to do extensive tests on other platforms/flavours.

> do you have any manual how to configure it?

currenlty not, but we welcome any contribution even guides like that.

**Comment by wajda** [comment]
> does Spline support AWS Glues

In short, yes, it does. Like Adam said, there could be some rough edges, and caveats, depending on your particular setup and job, but many people do use Spline on Glue successfully.

> do you have any manual how to configure it?

For example this - an older AWS article. 
https://aws.amazon.com/blogs/big-data/build-data-lineage-for-data-lakes-using-aws-glue-amazon-neptune-and-spline/

**Reply by zacayd** [reply]
thanks- I have read the article and put 
the following parameters on the Glue job

Dependent JARs path
path of  `spark-3.1-spline-agent-bundle_2.12-2.0.0.jar`

parameters:

```properties
spark.spline.mode=ENABLED
spark.spline.lineageDispatcher.http.producer.url=https://databricks.spline.octopai.com/producer
spark.spline.lineageDispatcher=http
spark.sql.queryExecutionListeners=za.co.absa.spline.harvester.listener.SplineQueryExecutionListener
```
also tried
```properties
spark.sql.queryExecutionListeners=za.co.absa.spline.harvester.listener.SplineQueryExecutionListener
spark.spline.producer.url=https://databricks.spline.octopai.com/producer
```
but didnt got any lineage on the arangoDB

the code of the job is simple- write from csv to other csv
```python
import sys
from awsglue.transforms import *
from awsglue.utils import getResolvedOptions
from pyspark.context import SparkContext
from awsglue.context import GlueContext
from awsglue.job import Job


args = getResolvedOptions(sys.argv, ["JOB_NAME"])

sc = SparkContext()
glueContext = GlueContext(sc)
spark = glueContext.spark_session
job = Job(glueContext)
job.init(args["JOB_NAME"], args)

# Read the CSV file from the source S3 location
source_path = "s3://johnny-chivers-glue-full-course-a/data/customers_source/Changes.csv"
df = spark.read.csv(source_path, header=True, inferSchema=True)

# Write the DataFrame to the target S3 location
target_path = "s3://johnny-chivers-glue-full-course-a/data/customers_source/ChangesTgt.csv"
df.write.csv(target_path, mode="overwrite", header=True)

# Commit the job
job.commit()
```
what am i missing? how to trouble shoot it?

**Comment by wajda** [comment]
Is there anything in logs?

**Comment by zacayd** [comment]
yes - the dirve logs
Driver logs
[Driver and executor log streams ](https://us-east-1.console.aws.amazon.com/cloudwatch/home#logStream:group=/aws-glue/jobs/logs-v2;prefix=jr_2933630a2a66fd301db83094a9e66a371205ba78d593c80a97af4249010c2dd7;streamFilter=typeLogStreamPrefix)

24/03/26 14:11:32 INFO ShutdownHookManager: Shutdown hook called
24/03/26 14:11:32 INFO DAGScheduler: Job 2 finished: csv at NativeMethodAccessorImpl.java:0, took 1.416374 s
24/03/26 14:11:32 INFO DAGScheduler: Job 2 is finished. Cancelling potential speculative or zombie tasks for this job
24/03/26 14:11:30 INFO DAGScheduler: Got job 2 (csv at NativeMethodAccessorImpl.java:0) with 1 output partitions
24/03/26 14:11:30 INFO DAGScheduler: Job 1 finished: csv at NativeMethodAccessorImpl.java:0, took 2.500862 s
24/03/26 14:11:30 INFO DAGScheduler: Job 1 is finished. Cancelling potential speculative or zombie tasks for this job
24/03/26 14:11:27 INFO DAGScheduler: Got job 1 (csv at NativeMethodAccessorImpl.java:0) with 1 output partitions
24/03/26 14:11:27 INFO DAGScheduler: Job 0 finished: csv at NativeMethodAccessorImpl.java:0, took 22.866388 s
24/03/26 14:11:27 INFO DAGScheduler: Job 0 is finished. Cancelling potential speculative or zombie tasks for this job
24/03/26 14:11:19 WARN TaskSchedulerImpl: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources
24/03/26 14:11:04 INFO DAGScheduler: Got job 0 (csv at NativeMethodAccessorImpl.java:0) with 1 output partitions
24/03/26 14:10:58 INFO GlueContext: GlueMetrics configured and enabled
24/03/26 14:10:55 INFO Utils: Successfully started service 'sparkDriver' on port 36727.

**Reply by wajda** [reply]
You have posted a link to a CloudWatch. I do not have access to your AWS account.

**Comment by zacayd** [comment]
I mean this is the log
24/03/26 14:11:32 INFO ShutdownHookManager: Shutdown hook called
24/03/26 14:11:32 INFO DAGScheduler: Job 2 finished: csv at NativeMethodAccessorImpl.java:0, took 1.416374 s
24/03/26 14:11:32 INFO DAGScheduler: Job 2 is finished. Cancelling potential speculative or zombie tasks for this job
24/03/26 14:11:30 INFO DAGScheduler: Got job 2 (csv at NativeMethodAccessorImpl.java:0) with 1 output partitions
24/03/26 14:11:30 INFO DAGScheduler: Job 1 finished: csv at NativeMethodAccessorImpl.java:0, took 2.500862 s
24/03/26 14:11:30 INFO DAGScheduler: Job 1 is finished. Cancelling potential speculative or zombie tasks for this job
24/03/26 14:11:27 INFO DAGScheduler: Got job 1 (csv at NativeMethodAccessorImpl.java:0) with 1 output partitions
24/03/26 14:11:27 INFO DAGScheduler: Job 0 finished: csv at NativeMethodAccessorImpl.java:0, took 22.866388 s
24/03/26 14:11:27 INFO DAGScheduler: Job 0 is finished. Cancelling potential speculative or zombie tasks for this job
24/03/26 14:11:19 WARN TaskSchedulerImpl: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources
24/03/26 14:11:04 INFO DAGScheduler: Got job 0 (csv at NativeMethodAccessorImpl.java:0) with 1 output partitions
24/03/26 14:10:58 INFO GlueContext: GlueMetrics configured and enabled
24/03/26 14:10:55 INFO Utils: Successfully started service 'sparkDriver' on port 36727.

**Reply by wajda** [reply]
I do not see any logs related to Spline. Please find the relevant logs. They must be among driver logs.
If there is none then the Spline wasn't installed for a job. In that case try using programmatic init method and see if it works.

**Reply by zacayd** [reply]
programicly means put
`sc._jvm.za.co.absa.spline.harvester.SparkLineageInitializer.enableLineageTracking(spark._jsparkSession)`
?

**Reply by wajda** [reply]
yes

**Reply by zacayd** [reply]
also not working
strange- when i add to the conf
--conf spark.sql.queryExecutionListeners=za.co.absa.spline.harvester.listener.SplineQueryExecutionListener
then it fails

**Comment by zacayd** [comment]
Hi seems to work when i put
--conf
spark.spline.producer.url=https://databricks.spline.octopai.com/producer --conf spark.sql.queryExecutionListeners=za.co.absa.spline.harvester.listener.SplineQueryExecutionListener
 
--extra-jars
--user-jars-first

Dependent JARs path
s3://glue-987771459238-us-east-1/lib/spark-3.1-spline-agent-bundle_2.12-0.6.1.jar

But has issues with spark 3.3

LAUNCH ERROR | Error downloading from S3 for bucket: johnny-chivers-glue-full-course-a, key: data/spark-3.3-spline-agent-bundle_2.12-2.0.0.jar;s3://johnny-chivers-glue-full-course-a/data/snakeyaml-1.33.jar.The specified key does not exist. (Service: Amazon S3; Status Code: 404; Please refer logs for details.

**Reply by wajda** [reply]
I believe you used an incorrect delimiter for your JAR URLs. But the issue doesn't seem to be related to Spline at this point.

**Reply by zacayd** [reply]
i get an error 
Is it familiar to you?
Error Category: UNCLASSIFIED_ERROR; An error occurred while calling o97.read. org/yaml/snakeyaml/Yaml

**Reply by wajda** [reply]
It might be caused by Glue 4.0 missing `snakeyaml` jar, so you should add it as a dependency for the job.
